internal/persist/slice.go

```diff
@@ -0,0 +1,545 @@
+	// As a special case, the tree of height 0 may have tree set to any(nil) or (*inode)(nil),
+	// and there are no trees of height 1 (which would need a root of type *leaf[T]).
+	tree   atomic.Value // tree of chunks
+	height int          // height of tree
+	tlen   int          // number of elements in tree (excludes tail)
+	tail   []T          // pending values (up to chunk) to flush to tree
+}
+
+// A TransientSlice is a mutable slice of values,
+// typically derived from and intended to become a [Slice].
+type TransientSlice[T any] struct {
+	// A TransientSlice is the mutable form of a Slice.
+	// It uses the same 16-way tree, but copy-on-write.
+	// More specifically, the TransientSlice keeps track
+	// of which nodes and leaves it owns, meaning they
+	// are not shared with any Slices. Those can be modified
+	// directly. Other nodes and leaves must be copied before
+	// being modified.
+	//
+	// Each TransientSlice has a unique ID (the aid field below, accessed by [TransientSlice.id]).
+	// A node or leaf with the same ID is owned by that TransientSlice
+	// and writable; others are copy-on-write.
+	// The [TransientSlice.Persist] operation picks a new top-level ID,
+	// giving up ownership of all the previously owned nodes and leaves
+	// and making them safe to publish as a Slice.
+	//
+	// In addition to the tree structure, a Slice has a tail of up to 16
+	// elements not yet stored in the tree, cutting tree updates by 16X.
+	// The wtail field tracks whether s.tail is owned by the TransientSlice,
+	// meaning can be written to. If wtail is true, s.tail has capacity 16.
+	//
+	// The tailLen is a cached copy of len(s.tail), necessary because
+	// [TransientSlice.writeTail] makes a copy of and reassigns s.tail.
+	// This might happen during t.Set(i, ...) and does not change len(s.tail),
+	// so conceptually t.Len() should be permitted concurrent with t.Set(i, ...),
+	// but technically reading len(s.tail) during t.Len and writing it during
+	// writeTail during t.Set is a race. Instead, we only set s.tail using
+	// t.setTail, which caches len(s.tail) in t.tailLen.
+	s       Slice[T]      // underlying slice data structure
+	aid     atomic.Uint64 // id of this transient (see [TransientSlice.id])
+	tailLen int           // cached copy of len(t.s.tail)
+	wtail   atomic.Bool   // whether s.tail is writable
+	tailMu  sync.Mutex    // protects tail copy in writeTail
+}
+
+// transientID is the most recently used transient ID.
+// To obtain a new ID, use transientID.Add(1).
+var transientID atomic.Uint64
+
+// id returns the id for this TransientSlice, copied from t.aid.
+// If an id has not been allocated yet (t.aid==0),
+// id sets t.aid to a new id and returns that id.
+func (t *TransientSlice[T]) id() uint64 {
+Retry:
+	id := t.aid.Load()
+	if id == 0 {
+		id = transientID.Add(1)
+		if !t.aid.CompareAndSwap(0, id) {
+			// Racing with another t.id call.
+			goto Retry
```

Comment:
no goto please. it's easy to write this as a loop.

Reply:
Done.